File: cxx/traaxx/sources/propagate_down_single_pass.cpp
#include <traaxx/propagate_down_single_pass.hpp>

namespace traaxx
{
    template<typename IndexT>
    BitMask propagate_down_single_pass(const std::vector<IndexT> &parent, BitMask seed)
    {
        auto const n = static_cast<IndexT>(parent.size());
        for (IndexT i = 0; i < n; ++i)
        {
            if (seed.get(parent[i]))
            {
                seed.set(i);
            }
        }
        return seed;
    }

    template BitMask propagate_down_single_pass<std::uint32_t>(const std::vector<std::uint32_t> &, BitMask);
}

File: go/std/propagate_down_single_pass.go
package std

// PropagateDownSinglePass propagates a mask downward in a single sequential
// pass, requiring the topological order invariant parent[i] <= i. Mirrors
// traaxx::propagate_down_single_pass
// (cxx/traaxx/sources/propagate_down_single_pass.cpp). Clones seed before
// mutating it, so — unlike a raw Go slice — the caller's mask is left
// untouched, matching the C++ pass-by-value seed parameter.
func PropagateDownSinglePass(parent []Index, seed BitMask) BitMask {
	result := seed.Clone()
	n := len(parent)
	for i := 0; i < n; i++ {
		if result.Get(parent[i]) {
			result.Set(Index(i))
		}
	}
	return result
}
